### PR TITLE
Support `run_id IN` and `run ID NOT IN` for `search_runs`

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## Related Issues/PRs

#6945

## What changes are proposed in this pull request?

Support `run_id IN` and `run ID NOT IN` for `search_runs`.

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/search_utils.py [hunk 1]
```diff
@@ -251,6 +257,14 @@ def _get_value(cls, identifier_type, key, token):
                 return token.value
             elif token.ttype in cls.STRING_VALUE_TYPES or isinstance(token, Identifier):
                 return cls._strip_quotes(token.value, expect_quoted_value=True)
+            elif isinstance(token, Parenthesis):
+                if key != "run_id":
+                    raise MlflowException(
+                        "Only run_id attribute support compare with a list of quoted string "
+                        "values.",
```
Comment (dbczumar) [inline mlflow/utils/search_utils.py:264]:
```suggestion
                        "Only the 'run_id' attribute supports comparison with a list of quoted string "
                        "values.",
```

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy !

**Comment by dbczumar** [comment]
@harupy Can we wait until @alexacole files a PR that introduces a test for run ID equality search before merging? https://github.com/mlflow/mlflow/issues/6867#issuecomment-1263930674

File: mlflow/utils/search_utils.py (hunk 1 above)
Comment (harupy) [reply mlflow/utils/search_utils.py]: Fixed!

**Comment by harupy** [comment]
@dbczumar Sure!

File: tests/utils/test_search_utils.py [hunk 2]
```diff
@@ -128,8 +128,6 @@ def test_correct_quote_trimming(filter_string, parsed_filter):
         ("`dummy.A > 0.1", "Invalid clause(s) in filter string"),
         ("dummy`.A > 0.1", "Invalid clause(s) in filter string"),
         ("attribute.start != 1", "Invalid attribute key"),
-        ("attribute.run_id != 1", "Invalid attribute key"),
-        ("attribute.run_uuid != 1", "Invalid attribute key"),
```
Comment (harupy) [inline tests/utils/test_search_utils.py:132]: @alexacole This change needs be imported in #6945

File: tests/utils/test_search_utils.py [hunk 3]
```diff
@@ -492,7 +490,6 @@ def test_order_by_metric_with_nans_infs_nones():
         ("`metrics.A", "Invalid order_by clause"),
         ("`metrics.A`", "Invalid entity type"),
         ("attribute.start", "Invalid attribute key"),
-        ("attribute.run_id", "Invalid attribute key"),
```
Comment (harupy) [inline tests/utils/test_search_utils.py:495]: @alexacole This change needs be imported in #6945

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 4]
```diff
@@ -2520,11 +2554,12 @@ def test_get_attribute_name():
     assert models.SqlRun.get_attribute_name("end_time") == "end_time"
     assert models.SqlRun.get_attribute_name("deleted_time") == "deleted_time"
     assert models.SqlRun.get_attribute_name("run_name") == "name"
+    assert models.SqlRun.get_attribute_name("run_id") == "run_uuid"
 
     # we want this to break if a searchable or orderable attribute has been added
     # and not referred to in this test
     # searchable attributes are also orderable
-    assert len(entities.RunInfo.get_orderable_attributes()) == 6
+    assert len(entities.RunInfo.get_orderable_attributes()) == 7
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:2562]: @alexacole This change needs be imported in #6945

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 5]
```diff
@@ -2520,11 +2554,12 @@ def test_get_attribute_name():
     assert models.SqlRun.get_attribute_name("end_time") == "end_time"
     assert models.SqlRun.get_attribute_name("deleted_time") == "deleted_time"
     assert models.SqlRun.get_attribute_name("run_name") == "name"
+    assert models.SqlRun.get_attribute_name("run_id") == "run_uuid"
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:2557]: @alexacole This change needs be imported in #6945

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 6]
```diff
@@ -1822,12 +1822,7 @@ def test_search_attrs(self):
         filter_string = "attribute.artifact_uri ILIKE '%{}%'".format(r1[-16:].upper())
         self.assertCountEqual([r1], self._search([e1, e2], filter_string))
 
-        for (k, v) in {
-            "experiment_id": e1,
-            "lifecycle_stage": "ACTIVE",
-            "run_id": r1,
-            "run_uuid": r2,
-        }.items():
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:1830]: @alexacole This change needs be imported in #6945

File: mlflow/store/tracking/dbmodels/models.py [hunk 7]
```diff
@@ -198,7 +198,10 @@ def get_attribute_name(mlflow_attribute_name):
         # Currently, MLflow Search attributes defined in `SearchUtils.VALID_SEARCH_ATTRIBUTE_KEYS`
         # share the same names as their corresponding `SqlRun` attributes. Therefore, this function
         # returns the same attribute name
-        return "name" if mlflow_attribute_name == "run_name" else mlflow_attribute_name
+        return {
+            "run_name": "name",
+            "run_id": "run_uuid",
+        }.get(mlflow_attribute_name, mlflow_attribute_name)
```
Comment (harupy) [inline mlflow/store/tracking/dbmodels/models.py:204]: @alexacole This change needs be imported in #6945

File: tests/store/tracking/test_file_store.py [hunk 8]
```diff
@@ -1063,6 +1063,46 @@ def test_search_runs_run_name(self):
...
+            run_view_type=ViewType.ACTIVE_ONLY,
+        )
+        assert [r.info.run_id for r in result] == [run_id1]
+
+        result = fs.search_runs(
+            [exp_id],
+            filter_string=f"attributes.run_id NOT IN ('{run_id1}')",
+            run_view_type=ViewType.ACTIVE_ONLY,
+        )
+        assert [r.info.run_id for r in result] == [run_id2]
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:1087]: @alexacole In https://github.com/mlflow/mlflow/pull/6945, we can add a test that looks like this:

```suggestion
    def test_search_runs_run_id(self):
        fs = FileStore(self.test_root)
        exp_id = fs.create_experiment("test_search_runs_run_id")
        # Set start_time to ensure the search result is deterministic
        run1 = fs.create_run(exp_id, user_id="user", start_time=1, tags=[], run_name="1")
        run2 = fs.create_run(exp_id, user_id="user", start_time=2, tags=[], run_name="2")
        run_id1 = run1.info.run_id
        run_id2 = run2.info.run_id

        result = fs.search_runs(
            [exp_id],
            filter_string=f"attributes.run_id = '{run_id1}'",
            run_view_type=ViewType.ACTIVE_ONLY,
        )
        assert [r.info.run_id for r in result] == [run_id1]

        result = fs.search_runs(
            [exp_id],
            filter_string=f"attributes.run_id != '{run_id1}'",
            run_view_type=ViewType.ACTIVE_ONLY,
        )
        assert [r.info.run_id for r in result] == [run_id2]
```

File: tests/store/tracking/test_sqlalchemy_store.py (hunk 5 above)
Comment (alexacole) [reply tests/store/tracking/test_sqlalchemy_store.py:2557]: @harupy Hello, I've tried going about updating my branch and fetching everything, but I'm still unable to import the changes in #6945

Comment (harupy) [reply tests/store/tracking/test_sqlalchemy_store.py:2557]: @alexacole Got it. You can manually add those changes to your PR.

**Pull request merged**